state/protocol/badger/validity.go

```diff
@@ -111,3 +117,161 @@ func validCommit(commit *flow.EpochCommit, setup *flow.EpochSetup) error {
 
 	return nil
 }
+
+type receiptValidator struct {
+	state    protocol.State
+	index    storage.Index
+	results  storage.ExecutionResults
+	verifier module.Verifier
+}
+
+func NewReceiptValidator(state protocol.State, index storage.Index, results storage.ExecutionResults) protocol.ReceiptValidator {
+	rv := &receiptValidator{
+		state:    state,
+		index:    index,
+		results:  results,
+		verifier: signature.NewAggregationVerifier(encoding.ExecutionReceiptTag),
+	}
+
+	return rv
+}
+
+// checkIsStakedNodeWithRole checks whether, at the given block, `nodeID`
+//   * is an authorized member of the network
+//   * has _positive_ weight
+//   * and has the expected role
+// Returns the following errors:
+//   * sentinel engine.InvalidInputError if any of the above-listed conditions are violated.
+//   * generic error indicating a fatal internal bug
+// Note: the method receives the block header as proof of its existence.
+// Therefore, we consider the case where the respective block is unknown to the
+// protocol state as a symptom of a fatal implementation bug.
+func (v *receiptValidator) ensureStakedNodeWithRole(nodeID flow.Identifier, identity *flow.Identity, expectedRole flow.Role) error {
+
+	// check that the origin is a verification node
+	if identity.Role != expectedRole {
+		return engine.NewInvalidInputErrorf("expected node %x to have identity %s but got %s", nodeID, expectedRole, identity.Role)
+	}
+
+	// check if the identity has a stake
+	if identity.Stake == 0 {
+		return engine.NewInvalidInputErrorf("node has zero stake (%x)", identity.NodeID)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) identityForNode(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+	// get the identity of the origin node
+	identity, err := v.state.AtBlockID(blockID).Identity(nodeID)
+	if err != nil {
+		if protocol.IsIdentityNotFound(err) {
+			return nil, engine.NewInvalidInputErrorf("unknown node identity: %w", err)
+		}
+		// unexpected exception
+		return nil, fmt.Errorf("failed to retrieve node identity: %w", err)
+	}
+
+	return identity, nil
+}
+
+func (v *receiptValidator) verifySignature(receipt *flow.ExecutionReceipt, nodeIdentity *flow.Identity) error {
+	id := receipt.ID()
+	valid, err := v.verifier.Verify(id[:], receipt.ExecutorSignature, nodeIdentity.StakingPubKey)
```

Comment:
⚠️ 
* signing the ID of the receipt does not work:
   - the receipt itself contains the creator's signature, which would create a circular dependency 
   - instead, the Execution Node would sign the ID of the `ExecutionReceipt.Body()`
* the execution node currently does not sign its receipts 😕 
https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/engine/execution/ingestion/engine.go#L1276 


Generally, I would suggest:
*  we make a separate issue for the cryptography; implementing it simultaneously in Execution and Consensus node
    -  Create dedicated structs for signing and verifying Execution Receipts in a central place, which are used by Execution node and other nodes; the implementation could follow [`SingleSigner ` and  `SingleSignerVerifier` for consensus messages](https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/consensus/hotstuff/verification/single_signer.go)

Reply:
I was checking [here](https://github.com/onflow/flow-go/blob/master/engine/execution/ingestion/engine.go#L1287) seems like we have a signature, but indeed I don't like how it's built at the moment.